###  Exploratory Data Analyse

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)

#### 1. Data set

In [2]:
df_South_asain = pd.read_csv('South_Asian_dataset.csv')
df_South_asain

Country  Year  GDP (current US$)  GDP growth (annual %)  \
0    Afghanistan  2000       3.521418e+09                    NaN   
1    Afghanistan  2001       2.813572e+09              -9.431974   
2    Afghanistan  2002       3.825701e+09              28.600001   
3    Afghanistan  2003       4.520947e+09               8.832278   
4    Afghanistan  2004       5.224897e+09               1.414118   
..           ...   ...                ...                    ...   
187     Pakistan  2019       3.209090e+11               2.497637   
188     Pakistan  2020       3.004260e+11              -1.274087   
189     Pakistan  2021       3.485170e+11               6.513886   
190     Pakistan  2022       3.747880e+11               4.767540   
191     Pakistan  2023       3.383680e+11              -0.004855   

     GDP per capita (current US$)  \
0                      180.188369   
1                      142.903364   
2                      182.174038   
3                      199.643226   
4                      221.830531   
..                            ...   
187                   1437.165833   
188                   1322.314785   
189                   1506.108293   
190                   1589.263980   
191                   1407.021351   

     Unemployment, total (% of total labor force) (modeled ILO estimate)  \
0                                                7.955                     
1                                                7.958                     
2                                                7.939                     
3                                                7.922                     
4                                                7.914                     
..                                                 ...                     
187                                              4.830                     
188                                              6.162                     
189                                              6.338                     
190                                              5.548                     
191                                              5.499                     

     Inflation, consumer prices (annual %)  \
0                                      NaN   
1                                      NaN   
2                                      NaN   
3                                      NaN   
4                                      NaN   
..                                     ...   
187                              10.578362   
188                               9.739993   
189                               9.496211   
190                              19.873860   
191                              30.768128   

     Foreign direct investment, net inflows (% of GDP)  Trade (% of GDP)  \
0                                             0.004828               NaN   
1                                             0.024169               NaN   
2                                             1.306950               NaN   
3                                             1.278493               NaN   
4                                             3.577104               NaN   
..                                                 ...               ...   
187                                           0.696146         28.905576   
188                                           0.684695         26.716280   
189                                           0.616039         27.030422   
190                                           0.390087         33.040373   
191                                           0.537284         28.102077   

     Gini index  Population, total  Population growth (annual %)  \
0           NaN           19542982                      1.443803   
1           NaN           19688632                      0.742517   
2           NaN           21000256                      6.449321   
3           NaN           22645130                      7.541019   
4           NaN           23

#### 2. check information

In [3]:
df_South_asain.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 192 entries, 0 to 191
Data columns (total 33 columns):
 #   Column                                                                      Non-Null Count  Dtype  
---  ------                                                                      --------------  -----  
 0   Country                                                                     192 non-null    object 
 1   Year                                                                        192 non-null    int64  
 2   GDP (current US$)                                                           190 non-null    float64
 3   GDP growth (annual %)                                                       189 non-null    float64
 4   GDP per capita (current US$)                                                190 non-null    float64
 5   Unemployment, total (% of total labor force) (modeled ILO estimate)         192 non-null    float64
 6   Inflation, consumer prices (annual %)             

#### 3. Check colums names

In [4]:
df_South_asain.columns

Index(['Country', 'Year', 'GDP (current US$)', 'GDP growth (annual %)',
       'GDP per capita (current US$)',
       'Unemployment, total (% of total labor force) (modeled ILO estimate)',
       'Inflation, consumer prices (annual %)',
       'Foreign direct investment, net inflows (% of GDP)', 'Trade (% of GDP)',
       'Gini index', 'Population, total', 'Population growth (annual %)',
       'Poverty headcount ratio at $2.15 a day (2017 PPP) (% of population)',
       'Life expectancy at birth, total (years)',
       'Mortality rate, infant (per 1,000 live births)',
       'Literacy rate, adult total (% of people ages 15 and above)',
       'School enrollment, primary (% gross)',
       'Urban population (% of total population)',
       'Access to electricity (% of population)',
       'People using at least basic drinking water services (% of population)',
       'People using at least basic sanitation services (% of population)',
       'Carbon dioxide (CO2) emissions excluding LU

#### 4. Details each featurs

- **Country:** The name of the South Asian country being analyzed (e.g., India, Pakistan, Bangladesh).
- **Year:** The calendar year for which the data is recorded.
- **GDP (current US\$)**: The total market value of all goods and services produced within a country in a given year, expressed in current U.S. dollars without adjusting for inflation.
- **GDP growth (annual \%):** The annual percentage change in the GDP, indicating how much the economy has grown or shrunk compared to the previous year.
- **GDP per capita (current US\$):** GDP divided by the total population, showing the average economic output per person in current U.S. dollars.
- **Unemployment, total (% of total labor force) (modeled ILO estimate):** The percentage of the labor force that is unemployed, based on estimates from the International Labour Organization.
- **Inflation, consumer prices (annual %):** The annual percentage change in consumer prices, reflecting the rate at which the cost of goods and services increases or decreases.
- **Foreign direct investment, net inflows (% of GDP):** The net amount of foreign investments flowing into the country as a percentage of its GDP.
- **Trade (% of GDP):** The total value of imports and exports as a percentage of GDP, indicating the country's trade openness.
- **Gini index:** A measure of income inequality within a country, where 0 represents perfect equality and 100 represents maximum inequality.
- **Population, total :** The total number of people living in the country.
- **Population growth (annual %):** The annual percentage change in the total population.
- **Poverty headcount ratio at \$2.15 a day (2017 PPP) (% of population):** The percentage of the population living on less than \$2.15 a day (adjusted for purchasing power parity, PPP), a measure of extreme poverty.
- **Life expectancy at birth, total (years):** The average number of years a newborn is expected to live, assuming current mortality rates.
- **Mortality rate, infant (per 1,000 live births):** The number of infant deaths per 1,000 live births in a given year.
- **Literacy rate, adult total (% of people ages 15 and above):** The percentage of people aged 15 and above who can read and write.
- **School enrollment, primary (% gross):** The total primary school enrollment as a percentage of the population eligible for primary education.
- **Urban population (% of total population):** The percentage of the total population living in urban areas.
- **Access to electricity (% of population):** The percentage of the population with access to electricity.
- **People using at least basic drinking water services (% of population):** The percentage of the population with access to at least basic water services.
- **People using at least basic sanitation services (% of population):** The percentage of the population with access to basic sanitation facilities.
- **Carbon dioxide (CO2) emissions excluding LULUCF per capita (t CO2e/capita):** The per capita CO2 emissions, excluding land use, land-use change, and forestry (LULUCF).
- **PM2.5 air pollution, mean annual exposure (micrograms per cubic meter):** The mean annual concentration of fine particulate matter (PM2.5) in the air, measured in micrograms per cubic meter.
- **Renewable energy consumption (% of total final energy consumption):** The percentage of total energy consumed that comes from renewable sources.
- **Forest area (% of land area):** The percentage of the country’s total land area covered by forests.
- **Control of Corruption: Percentile Rank:** A governance indicator measuring perceptions of corruption in the public sector, ranked as a percentile.
- **Political Stability and Absence of Violence/Terrorism: Estimate:** A governance indicator reflecting the likelihood of political instability and/or violence, such as terrorism.
- **Regulatory Quality: Estimate:** A governance indicator that measures the ability of the government to formulate and implement sound policies.
- **Rule of Law: Estimate:** A governance indicator measuring the extent to which agents have confidence in and abide by societal rules, including contract enforcement, property rights, and the judiciary.
- **Voice and Accountability: Estimate:** A governance indicator that reflects the degree to which citizens can participate in  selecting their government and the extent of freedom of expression, association, and media.
- **Individuals using the Internet (% of population):** The percentage of the population with access to the internet.
- **Research and development expenditure (% of GDP):** The percentage of GDP spent on research and development activities.
- **High-technology exports (% of manufactured exports):** The share of high-tech products in the country's total manufactured exports, indicating technological advancement.

#### 5. Check missing Values:

In [5]:
df_South_asain.isnull().sum() 

Country                                                                         0
Year                                                                            0
GDP (current US$)                                                               2
GDP growth (annual %)                                                           3
GDP per capita (current US$)                                                    2
Unemployment, total (% of total labor force) (modeled ILO estimate)             0
Inflation, consumer prices (annual %)                                           9
Foreign direct investment, net inflows (% of GDP)                               5
Trade (% of GDP)                                                               51
Gini index                                                                    150
Population, total                                                               0
Population growth (annual %)                                                    0
Poverty headcoun

#### 6. Handle missing values and Check missing values:
#### 6.1.  Replace placeholders ('..') with NaN and convert columns to appropriate types

In [7]:
df_South_asain.replace('..', pd.NA, inplace=True)

# Convert all applicable columns to numeric:
for column in df_South_asain.columns:
    try:
        df_South_asain[column] = pd.to_numeric(df_South_asain[column], errors='ignore')
    except:
        pass

missing_values_summary = df_South_asain.isnull().sum().sort_values(ascending=False)
missing_values_summary[missing_values_summary > 0]

Poverty headcount ratio at $2.15 a day (2017 PPP) (% of population)           150
Gini index                                                                    150
Research and development expenditure (% of GDP)                               144
Literacy rate, adult total (% of people ages 15 and above)                    131
High-technology exports (% of manufactured exports)                           116
Trade (% of GDP)                                                               51
School enrollment, primary (% gross)                                           26
PM2.5 air pollution, mean annual exposure (micrograms per cubic meter)         24
Individuals using the Internet (% of population)                               19
Voice and Accountability: Estimate                                             16
Rule of Law: Estimate                                                          16
Regulatory Quality: Estimate                                                   16
Political Stabil

In [9]:
# Drop columns with more than 50% missing data
threshold = 0.5 
columns_to_drop = missing_values_summary[missing_values_summary > len(df_South_asain) * threshold].index
df_cleaned = df_South_asain.drop(columns=columns_to_drop)

# Impute missing values 
numeric_columns = df_cleaned.select_dtypes(include=['number']).columns
df_cleaned[numeric_columns] = df_cleaned[numeric_columns].fillna(df_cleaned[numeric_columns].median())

# Confirm no missing values remain in numeric columns
missing_after_imputation = df_cleaned.isnull().sum().sort_values(ascending=False)
missing_after_imputation[missing_after_imputation > 0]


Series([], dtype: int64)